### PR TITLE
Add support for pnpm lock file format 6

### PR DESCRIPTION
**Pull request by CraigMacomber**
This addresses https://github.com/microsoft/component-detection/issues/503 by doing the following:

- Explicitly mark code specific to version 5 as "V5"
- Add new code for V6.
- Add version detection logic to select which code to run.
- Add tests for new code.

The old code wasn't clear about what file formats it actually worked with, and a lot of its tests seems to be from versions earlier than 5 which it didn't really parse correctly (ex: the version it parses isn't present in those versions as it had a different name).
Thus to preserve existing behavior I pointed cases which failed to detect a version through the old V5 code.

The included verification tests didn't cover a lot of cases (link and file references, aliased packages, shared shrink-wrap format, disambiguated deps with peer deps) so I also tested on the package lock from https://github.com/microsoft/FluidFramework which is 44k lines and has all the mentioned edge cases.
If desired, I can add the v5 and v6 versions of that file to the verification tests directory.

This should be in a working state but I need to actually understand how the verification tests results are verifies before this should be merged.

This is my first time working in this codebase, as well as the first time in several years that I'm working in C#, so please be extra careful reviewing this: don't trust that I know what I'm doing.

### CODE REVIEW

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs [hunk 1]
```diff
@@ -104,14 +125,97 @@ private void RecordDependencyGraphFromFile(PnpmYaml yaml, ISingleFileComponentRe
...
+        foreach (var (_, (component, package)) in components)
+        {
+            foreach (var (name, version) in package.dependencies ?? Enumerable.Empty<KeyValuePair<string, string>>())
+            {
+                var pnpmDependencyPath = PnpmParsingUtilities.ReconstructPnpmDependencyPathV6(name, version);
+
+                // If this lookup fails, then pnpmDependencyPath was either parsed incorrectly or constructed incorrectly.
+                var (referenced, _) = components[pnpmDependencyPath];
+
+                singleFileComponentRecorder.RegisterUsage(referenced, parentComponentId: component.Component.Id, isExplicitReferencedDependency: false);
```
Comment (CraigMacomber) [inline src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs:167]: Its unclear to me if I need to respecify dev-dependency-ness here. Would doing so indicate this dependency edge is a dev dependency, or and I overwriting the dev-dependency ness of the target component with the default by not specifying it here?

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs [hunk 2]
```diff
@@ -30,7 +31,7 @@ public class PnpmComponentDetector : FileComponentDetector
 
     public override IEnumerable<ComponentType> SupportedComponentTypes { get; } = new[] { ComponentType.Npm };
 
-    public override int Version { get; } = 5;
+    public override int Version { get; } = 6;
```
Comment (CraigMacomber) [inline src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs:34]: I think I'm supposed to update this because I made changes, but I'd like confirmation that that's the correct thing to do.

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmParsingUtilities.cs [hunk 3]
```diff
@@ -3,43 +3,52 @@
...
-            throw new ArgumentNullException(nameof(pnpmPackage));
-        }
+        var deserializer = new DeserializerBuilder()
+            .IgnoreUnmatchedProperties()
+            .Build();
+        return deserializer.Deserialize<PnpmYamlV6>(new StringReader(fileContent));
+    }
 
-        return string.Equals(bool.TrueString, pnpmPackage.dev, StringComparison.InvariantCultureIgnoreCase);
+    /// <summary>
```
Comment (CraigMacomber) [inline src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmParsingUtilities.cs:37]: Nothing in here had doc comments before, but it seems like docs would be good so I added some. he projects configuration seems to like this format so I used it, but I'm not sure what the policy on documentation is.

**Comment by CraigMacomber** [comment]
It appears that the VS code dev container setup doesn't include PowerShell so I can't run the VerificationTests.ps1 script locally. The instructions in there also make explicit reference to visual studio. It also says "The automation script above will output variables that needs to be replaced in `ComponentDetectionIntegrationTests.cs` file.": as far as I can tell thats not accurate as it looks like that file pulls in the data from environment variables.

Anyway its unclear to me what the workflow is for adding more test cases to the verification script. I'd like to include my added v6 file works but I don't know how.

**Comment by CraigMacomber** [comment]
Looking at the code coverage, I determined I could use a test for the shared-shrinkwrap case that workspaces use, and one for renamed/aliased dependencies, so I have added both. More testing is possible (ex: file and link cases, peer deps etc.) but I think I'm past the point of diminishing returns.

**Comment by CraigMacomber** [comment]
@microsoft/ose-component-detection-maintainers : I'd like assistance dealing with the snapshot tests as noted in https://github.com/microsoft/component-detection/pull/1028#issuecomment-1992470325. If this is not practical to get assistance on, I can revert the extended snapshot test coverage which should get this passing on CI: maybe that would get this reviewed sooner?

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs (hunk 1 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs:167]: Our component merging logic will do the following where if any detection is classified as non-dev it will categorize the component as non-dev.

First Detection Dev Dep | Second Detection Dev Dep | Final Result
false                             | false                              | false
true                              | true                               | true
true                              | false                              | false

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs (hunk 2 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs:34]: I'd be happy to help make these changes as well if needed

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs (hunk 1 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs:167]: When passing null for isDevelopmentDependency whatever the previous value is will stay.

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs [hunk 4]
```diff
@@ -104,14 +125,97 @@ private void RecordDependencyGraphFromFile(PnpmYaml yaml, ISingleFileComponentRe
...
+        // "dedicated shrinkwrap" (single package) case:
+        ProcessDependencySet(yaml);
+
+        // "shared shrinkwrap" (workspace / mono-repos) case:
+        foreach (var (_, package) in yaml.importers ?? Enumerable.Empty<KeyValuePair<string, PnpmHasDependenciesV6>>())
+        {
+            ProcessDependencySet(package);
+        }
+
+        void ProcessDependencySet(PnpmHasDependenciesV6 item)
```
Comment (cobya) [inline src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs:182]: nit: can these be declared as explicit functions rather than locals for RecordDependencyGraphFromFileV6?

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmParsingUtilities.cs (hunk 3 above)
Comment (cobya) [reply src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmParsingUtilities.cs:47]: We appreciate any new docs being added, and yes this is the format we use 😄

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmParsingUtilities.cs [hunk 5]
```diff
@@ -50,7 +69,65 @@ private static (string Name, string Version) ExtractNameAndVersionFromPnpmPackag
...
+    {
+        /*
+         * The format is documented at https://github.com/pnpm/spec/blob/master/dependency-path.md.
+         * At the writing it does not seem to reflect changes which were made in lock file format v6:
+         * See https://github.com/pnpm/spec/issues/5.
+         */
+
+        // Strip parenthesized suffices from package. These hold peed dep related information that is unneeded here.
+        // An example of a dependency path with these: /webpack-cli@4.10.0(webpack-bundle-analyzer@4.10.1)(webpack-dev-server@4.6.0)(webpack@5.89.0)
+        var fullPackageNameAndVersion = pnpmDependencyPath.Split("(")[0];
```
Comment (cobya) [inline src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmParsingUtilities.cs:87]: is this guaranteed to always have [0]? do we need to do index checks beforehand?

**Comment by cobya** [comment]
> @microsoft/ose-component-detection-maintainers : I'd like assistance dealing with the snapshot tests as noted in [#1028 (comment)](https://github.com/microsoft/component-detection/pull/1028#issuecomment-1992470325). If this is not practical to get assistance on, I can revert the extended snapshot test coverage which should get this passing on CI: maybe that would get this reviewed sooner?

When adding new components like this, the verification test failures are expected as it is performing a diff based on the last "known good version" so adding new files will trip the failure. I'll verify before merging but you shouldn't need to worry about this for now.

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs (hunk 2 above)
Comment (CraigMacomber) [reply src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs:34]: That sounds like an excellent way to reduce risk in cases like this.

> I'd be happy to help make these changes as well if needed

That would be great. I don't know this codebase well (I just spent a day and a half making this as part of a hackathon), so I suspect you could do that much more efficiently than I could. I'm happy to have you take over any part or all of this work.

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs (hunk 4 above)
Comment (CraigMacomber) [reply src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs:182]: I choose to do it this way so they could be scoped to v6 specific code, and not require as many parameters. They could of course be moved to a larger scope, given more specific names, and have everything passed in: that just seemed messier to me. If you refactor the v5 and v6 detector to be more separated, moving these to a larger scope would be more attractive.

File: src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmParsingUtilities.cs (hunk 5 above)
Comment (CraigMacomber) [reply src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmParsingUtilities.cs:87]: My understanding is splitting a string always returns at least one part, even if there is no separator or the string is empty, so this should be fine. There likely is a more explicit way to do this (for example with a regular expression), but this should work fine.

**Comment by cobya** [comment]
Closing this as #1110 has been opened with the related experiments changes.
